net/wireguard/src/www/widgets/widgets/wireguard.widget.php

```diff
@@ -30,50 +31,93 @@
 require_once("guiconfig.inc");
 require_once("widgets/include/wireguard.inc");
 
-$data = trim(configd_run('wireguard showhandshake'));
+$enabled = ($config["OPNsense"]["wireguard"]["general"]["enabled"] === "1" ? true : false);
 
-$empty = strlen($data) == 0;
 ?>
 
 <table class="table table-striped table-condensed">
     <thead>
         <tr>
+            <th><?= gettext("Name") ?></th>
             <th><?= gettext("Interface") ?></th>
             <th><?= gettext("Endpoint") ?></th>
             <th><?= gettext("Latest Handshake") ?></th>
         </tr>
     </thead>
-    <tbody>
+    <tbody id="wg-table-tbody">
 
-<?php if (!$empty):
-    $handshakes = explode("\n", $data);
+    <?php if (!$enabled): ?>
+    <tr>
+        <td colspan="4"><?= gettext("No WireGuard instance defined or enabled.") ?></td>
+    </tr>
+    <?php endif; ?>
 
-    foreach ($handshakes as $handshake):
-        $item = explode("\t", $handshake);
+    </tbody>
+</table>
 
-        $epoch = $item[2];
-        $latest = "-";
-        if ($epoch > 0):
-            $dt = new DateTime("@$epoch");
-            $dt->setTimezone(new DateTimeZone(date_default_timezone_get()));
-            $latest = $dt->format(gettext("Y-m-d H:i:sP"));
-        endif; ?>
+<script>
+$(window).on("load", function() {
+    function wgGenerateRow(name, interface, peerName, latestHandshake, status)
+    {
 
-    <tr>
-        <td><?= $item[0] ?></td>
-        <td><?= gettext(substr($item[1], 0, 10)) ?>...</td>
-        <td><?= $latest ?></td>
-    </tr>
+        // More user friendly naming
+        if (latestHandshake === "0000-00-00 00:00:00+00:00") {
```

Comment:
Any reason we hardcore this string here? Typically the controller takes care of this already.

Reply:
Because the time defaults to the 0 string any case there was no handshake done yet. I want the API to always report a date for consistency, instead of such a user-facing string.

Reply:
In all honesty, I don't think that user-facing translation strings based on the OPNsense system language belong in a API consumes by client-facing code. If the API is consumed externally, receiving "Never connected." in any of the possible language set server-side on OPNsense is not feasible to parse and process. (for example by 3rd party code like monitoring tools which could use the API)

Controller are responsible to receive, send or process the data. The client-side part to make prepare them user-friendly - that's the goal of this small IF-condition here.

It's the nature of having getting API to send machine-parseable data in a pre-defined way, and preparing it in a user-friendly way on the client-side depending on the output from the API. When one of the components (view or controller) is modified, the code authors responsibility is to take care not breaking things. That's something which needs to be validated before approving the PR.

In the end, the worst possible case is that customer sees `0000-00-00 00:00:00+00:00` here.

For example the gateway widget also does status-string comparisons on the client-side, which could break the logic if those hardcoded status value change: https://github.com/opnsense/core/blob/e58d7de7d4090524aa7e6f09900b77c43dbd820f/src/www/widgets/widgets/gateways.widget.php#L85

Also here the uptime in seconds is expected from the API and days calculated based on that, where user-friendly text is shown accordingly. If sometimes ever uptime is sent as minutes and changed server-side without changing JavaScript, it will break the logic as well: https://github.com/opnsense/core/blob/2fdaab15529dcf1fc4581e399e87f919c4b364a8/src/www/widgets/widgets/system_information.widget.php#L79-L83

That widget also performs string comparison form a value to "NO" and renders it in a user-friendly way: https://github.com/opnsense/plugins/blob/3bcbadaf1c84465915d35fb40b21aa187edb7593/sysutils/apcupsd/src/www/widgets/widgets/apcupsd.widget.php#L152-L155

If you really don't like that if-condition, I think I'm going to chop that one, as I don't see any better way to handle this.

Reply:
> If you really don't like that if-condition, I think I'm going to chop that one, as I don't see any better way to handle this.

That's fine by me. We can reassess later as the question comes up.

Reply:
ping ^^^